File: plugins/plugin-auto-trader/rust/src/actions/stop_trading.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::service::TradingService;
use crate::{Action, ActionExample, ActionResult};

pub struct StopTradingAction;

#[async_trait]
impl Action for StopTradingAction {
    fn name(&self) -> &str {
        "STOP_TRADING"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "END_TRADING",
            "STOP_AUTO_TRADING",
            "DISABLE_TRADING",
            "TURN_OFF_TRADING",
        ]
    }

    fn description(&self) -> &str {
        "Stop automated trading."
    }

    async fn validate(&self, message: &Value, _state: &Value) -> bool {
        let text = message
            .pointer("/content/text")
            .and_then(|v| v.as_str())
            .unwrap_or("");
        let lower = text.to_lowercase();
        lower.contains("stop") || lower.contains("end") || lower.contains("disable")
    }

    async fn handler(
        &self,
        _message: &Value,
        _state: &Value,
        service: Option<&TradingService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "TradingService is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        match svc.stop_trading().await {
            Ok(()) => ActionResult {
                success: true,
                text: "Auto-trading stopped.".to_string(),
                data: None,
                error: None,
            },
            Err(e) => ActionResult {
                success: false,
                text: e,
                data: None,
                error: Some("stop_failed".to_string()),
            },
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "Stop trading".to_string(),
            agent_response: "Auto-trading stopped.".to_string(),
        }]
    }
}

File: plugins/plugin-auto-trader/python/elizaos_plugin_auto_trader/actions/stop_trading.py
from __future__ import annotations

from dataclasses import dataclass

from elizaos_plugin_auto_trader.actions.common import ActionResult, Message
from elizaos_plugin_auto_trader.service import TradingService


@dataclass
class StopTradingAction:
    @property
    def name(self) -> str:
        return "STOP_TRADING"

    async def validate(self, message: Message, _state: dict) -> bool:  # type: ignore[type-arg]
        text = ((message.get("content") or {}).get("text") or "").lower()
        return "stop" in text or "end" in text or "disable" in text

    async def handler(
        self,
        message: Message,
        _state: dict,  # type: ignore[type-arg]
        service: TradingService | None = None,
    ) -> ActionResult:
        if service is None:
            return ActionResult(False, "TradingService is not available.", "missing_service")

        try:
            await service.stop_trading()
        except RuntimeError as exc:
            return ActionResult(False, str(exc), "stop_failed")

        return ActionResult(True, "Auto-trading stopped.")
